src/lib/ProjectM36/Client/Transaction.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module ProjectM36.Client.Transaction
+  ( Transaction
+  , TransactionError(..)
+  , TransactionCanceled(..)
+  , TransactionCancelReasonUnknown(..)
+  , runTransactionAtHead
+  , runTransactionWithRebase
+  , execute, execute'
+  , executeIO, executeIO'
+  , query, query'
+  , rollback, rollbackWith, maybeRollback, eitherRollback
+  ) where
+
+import           Control.Exception    (Exception, SomeException, catch, throwIO,
+                                       toException)
+import           Control.Monad.Reader
+import           Data.Typeable        (Typeable)
+import           ProjectM36.Base      (DatabaseContextIOExpr, Relation)
+import           ProjectM36.Client    hiding (rollback)
+import           ProjectM36.Error     (MergeError (DisconnectedTransactionNotAMergeHeadError),
+                                       RelationalError (MergeTransactionError))
+
+newtype Transaction a = Transaction (ReaderT (Connection, SessionId) IO a)
+  deriving (Functor, Applicative, Monad)
+
+data TransactionError = TransactionRelationalError RelationalError
+                      | TransactionCanceledError SomeException
+                      | TransactionRetriesExhausted
+                      deriving (Show, Typeable)
+instance Exception TransactionError
+
+newtype TransactionCanceled = TransactionCanceled SomeException deriving (Show, Typeable)
+instance Exception TransactionCanceled
+
+data TransactionCancelReasonUnknown = TransactionCancelReasonUnknown deriving (Show, Typeable)
+instance Exception TransactionCancelReasonUnknown
+
+
+runTransactionAtHead :: HeadName -> Connection -> Transaction a -> IO (Either TransactionError a)
+runTransactionAtHead branch conn trx = do
+  errSess <- createSessionAtHead branch conn
+  case errSess of
+    Left err -> pure $ Left (TransactionRelationalError err)
+    Right sess -> do
+        a <- addTransactionToSession sess conn trx
+        maybe (Right a) (Left . TransactionRelationalError) <$> commit sess conn
+      `catch` \(TransactionCanceled e) -> pure (Left $ TransactionCanceledError e)
+
+runTransactionWithRebase :: Int -> HeadName -> Connection -> Transaction a -> IO (Either TransactionError a)
+runTransactionWithRebase tries branch conn trx
+  | tries <= 0 = pure (Left TransactionRetriesExhausted)
+  | otherwise = do
+    errResult <- runTransactionAtHead branch conn trx
+    case errResult of
+      Left err -> case err of
+        (TransactionRelationalError (MergeTransactionError _)) -> runTransactionWithRebase (tries - 1) branch conn trx
```

Comment:
I'm not sure if I agree with this strategy. If there is a merge conflict, it seems unlikely that waiting for other commits to jump in between would resolve the conflict. I suspect, but I'm not certain, that this would be better accomplished server-side where a genuine determination of merge-ability could drive the rebase mechanism. It could be tricky, though- we don't currently store the expressions needed to move from one state to the next (like a git diff). I plan to change that because I need it for my lazy transactions patch.